gprofiler/spark/sampler.py

```diff
@@ -259,30 +66,26 @@ def _collect_loop(self) -> None:
         timefn = time.monotonic
         start_time = timefn()
         while not self._stop_event.is_set():
-            if self._spark_sampler is None:
-                spark_cluster_conf = self._find_spark_cluster()
-                if spark_cluster_conf is not None:
-                    master_address, cluster_mode = spark_cluster_conf
-                    self._spark_sampler = SparkCollector(cluster_mode, master_address)
-                else:
+            if self._spark_sampler is not None:
+                discovered = self._spark_sampler.discover()
+                if not discovered:
                     if timefn() - start_time >= FIND_CLUSTER_TIMEOUT_SECS:
                         logger.info("Timed out identifying Spark cluster. Stopping Spark collector.")
                         break
 
-            if self._spark_sampler is not None:
-                collected = self._spark_sampler.collect()
-                # No need to submit samples that don't actually have a value:
-                samples = tuple(filter(lambda s: s.value is not None, collected))
-                snapshot = MetricsSnapshot(datetime.now(tz=timezone.utc), samples)
-                if self._storage_dir is not None:
-                    now = get_iso8601_format_time(datetime.now())
-                    base_filename = os.path.join(self._storage_dir, f"spark_metric_{escape_filename(now)}")
-                    with open(base_filename, "w") as f:
-                        json.dump(bake_metrics_payload(snapshot), f)
-                if self._client is not None:
-                    self._client.submit_spark_metrics(snapshot)
-
-            self._stop_event.wait(self._sample_period)
+                elif discovered:
+                    snapshot = self._spark_sampler.snapshot()
+                    logger.debug("Collected Spark metrics", snapshot=snapshot)
+                    # No need to submit samples that don't actually have a value:
```

Comment:
Can `snapshot` be None here? It can't happen if discovery passed, can it? Then don't check it and just `assert snapshot is not None`, or alternatively, in the `snapshot` value, you can `assert` that it's called only for discovered clusters, and not if `discover` failed.

Also this code doesn't make much sense - you refrain from writing locally if it's None, but not refraining from submitting the metrics?

Reply:
https://github.com/Granulate/gprofiler/pull/749/files#diff-18dce65a9628cdb97bacc94da9b0401f8686d3326c5a71c91a23eb0bf4cbb919R78